### PR TITLE
feat: add migration start step

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Add migration start step
- If the expected leader region has been opened on `to_peer`, go to the `MigrationEnd` state.
- If the candidate region has been opened on `to_peer`, go to the `DowngradeLeader` state.
- Else go to the `OpenCandidateRegion` state.

**Behaviors:**

 Abort(non-retry):
 - TableRoute is not found.
 - RegionRoute is not found.
 - Leader peer of RegionRoute is not found.

 Retry:
 - Failed to retrieve the metadata of table.
## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#2700

### CODE REVIEW

**Comment by WenyXu** [comment]
@MichaelScofield @fengjiachun PTAL

File: src/meta-srv/src/procedure/region_migration.rs [hunk 1]
```diff
@@ -59,74 +69,112 @@ impl PersistentContext {
...
+        }
+    }
+}
+
 /// The context of procedure execution.
-#[derive(Debug, Clone)]
-pub struct Context {}
+pub struct Context {
+    persistent_ctx: Arc<Mutex<PersistentContext>>,
+    volatile_ctx: Mutex<VolatileContext>,
```
Comment (fengjiachun) [inline src/meta-srv/src/procedure/region_migration.rs:96]: It seems unnecessary to lock.

File: src/meta-srv/src/procedure/region_migration.rs [hunk 2]
```diff
@@ -59,74 +69,112 @@ impl PersistentContext {
...
+            table_metadata_manager: self.table_metadata_manager,
+        }
+    }
+}
+
 /// The context of procedure execution.
-#[derive(Debug, Clone)]
-pub struct Context {}
+pub struct Context {
+    persistent_ctx: Arc<Mutex<PersistentContext>>,
```
Comment (fengjiachun) [inline src/meta-srv/src/procedure/region_migration.rs:95]: It seems unnecessary to lock.

File: src/meta-srv/src/procedure/region_migration/migration_start.rs [hunk 3]
```diff
@@ -12,24 +12,302 @@
...
+impl RegionMigrationStart {
+    /// Retrieves region route.
+    ///
+    /// Abort(non-retry):
+    /// - TableRoute is not found.
+    /// - RegionRoute is not found.
+    ///
+    /// Retry:
+    /// - Failed to retrieve the metadata of table.
+    async fn retrieve_regions_route(
```
Comment (fengjiachun) [inline src/meta-srv/src/procedure/region_migration/migration_start.rs:73]: retrieve region route?

**Review by MichaelScofield — APPROVED** [review]

**Comment by WenyXu** [comment]
@fengjiachun PTAL

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**